feat: migrate security records console to tgui

BartDrown opened the pull request:
# About the pull request

Purpose of this PR is to refactor and modernize security records console code.

Security records console (and other records consoles overall) have been not updated properly for long time and contain a lot of technical debt. A lot of functionalities were not intuitive, didn't work properly or didn't work at all.

This PR tries to fix that and basically changes most of the security console code, at the same time keeping core functionalities working the same/as they should.

I've tested my changes locally, but testing under higher load would help ensure stability.

# Explain why it's good for the game

Core of the console functionalities didn't change much, but it's quality of life improvements.

Players are now able to use modern UI while accessing security records console.
It provides better user experience, handles states and inputs in modern way.

Records fields updates and validation are now more generic and extendable.
It creates good base for future features, as code is now easily manageable and maintainable.

# Testing Photographs and Procedure
<details>
<summary>Screenshots & Videos</summary>

Reworked Security Record view
![image](https://github.com/user-attachments/assets/9ac9d0af-4519-440e-9980-540f05888e56)

Reworked Security Records list
![image](https://github.com/user-attachments/assets/a9c39b28-11ce-4e16-bba8-71d41166bda2)

Reworked Fingerprint Scanner view
![image](https://github.com/user-attachments/assets/f886756f-f97b-4da7-baed-97ccf2eb82c7)

Assigned icons for fingerprints and security records
![image](https://github.com/user-attachments/assets/d6e13f95-19f2-48a4-926f-3ad5be087dc3)

Custom mugshot photo interface
![image](https://github.com/user-attachments/assets/64745745-c240-4329-9ad0-bc806beb48f8)

</details>


# Changelog
:cl:
add: migrate security records console to tgui
fix: highcom cannot change rank in record
fix: photo in security records console does not show
fix: photo in security records console cannot be updated
fix: fingerprint scanner report cannot be viewed if clue detector is not ejected and cleared
qol: add focus and listen on enter in inputs of security console
qol: add cursor pointer to TGUI Button component
refactor: refactor security records console code
code: security records client side sorting and filtering
/:cl:

There are many more, but I don't want to bloat this PR changelog

spookydonut commented on code/game/machinery/computer/security.dm:
```diff
@@ -34,12 +22,89 @@
 
 	. = ..()
 
+/obj/structure/machinery/computer/secure_data/ui_static_data(mob/user)
```
ui_static_data is only for data that doesn't change

BartDrown replied:
Changed it to ui_data. I was kinda worried it can be updated unnecessarily so that's why I used it before.

spookydonut commented on code/game/machinery/computer/security.dm:
```diff
@@ -34,12 +22,89 @@
+	// Map security records via id
+	var/list/records = list()
+	var/list/security_map = list()
+	for (var/datum/data/record/security in GLOB.data_core.security)
+		security_map[security.fields["id"]] = security
+
+	for (var/datum/data/record/general in GLOB.data_core.general)
```
im pretty sure datacore gets added to during the round as people spawn in after round start

spookydonut commented on code/game/machinery/computer/security.dm:
```diff
@@ -34,12 +22,89 @@
+
+	// Define criminal statuses and colors
+	var/list/criminal_statuses = list(
+	"*Arrest*" = list("background" = "#990c28", "font" = "#ffffff"),
+	"Incarcerated" = list("background" = "#faa20a", "font" = "#ffffff"),
+	"Released" = list("background" = "#2981b3", "font" = "#ffffff"),
+	"Suspect" = list("background" = "#686A6C", "font" = "#ffffff"),
+	"NJP" = list("background" = "#b60afa", "font" = "#ffffff"),
+	"None" = list("background" = "inherit", "font" = "inherit"),
```
style  stuff like this should be in the ui not here

BartDrown replied:
Moved this to UI, it wasn't needed here after all cause validation is still later in place.

spookydonut commented on code/game/machinery/computer/security.dm:
```diff
@@ -34,12 +22,89 @@
+/obj/structure/machinery/computer/secure_data/ui_static_data(mob/user)
+	. = ..()
+
+	if(!allowed(usr))
+		to_chat(user, SPAN_WARNING("Access denied."))
+		return
+
+	// Pass player data, only for frontend state
+	var/mob/living/carbon/human/U = usr
+	var/user_data = list("name" = U.get_authentification_name(), "rank" = U.get_assignment())
+	.["user_data"] = user_data
```
access and user data dont belong here, use ui_act to double check a user can do an action and pass anything about a current user with ui_data

BartDrown replied:
UI open and ui_act already checked for permission, it was redundant here. I removed user_data as well, cause it wasn't needed in UI anyway.

BartDrown commented:
Addressed all issues raised in the review, please re-review it now.

harryob commented on tgui/packages/tgui/interfaces/SecurityRecords.jsx:
for future reference, typescript interfaces are **much** preferred and make the linters work a hell of a lot better, i won't make you convert it but

BartDrown replied:
Okay, sure. I had enough problem trying to figure out state manegement and data transfer on it's own, that's why I decided to drop TS for now. I'll keep that in mind in the future.

harryob commented on tgui/packages/tgui/interfaces/SecurityRecords.jsx:
```diff
@@ -0,0 +1,688 @@
+import { Window } from '../layouts';
+
+export const SecurityRecords = () => {
+  // Le Funny hack cause browse_rsc doesn't get reloaded due to how browser works
+  // This adds dynamic parameter to GET request so resource must be refetch
+  // Probably could be improved but for now I have no idea how
```
ideally we do not rely on browse_rsc behavior at all in tgui, this should all be handled with SSassets

BartDrown replied:
Converted all `browser_rsc` to `icon2html`. I'am still not super happy about current implementation, cause now it's converting photo for every record at initialization. I'm worried about performance on server full of people.
If you could give me a hint how it could be further improved or if it's enough as is, I would highly appreciate it.

```dm
for (var/datum/data/record/general in GLOB.data_core.general)
	var/id = general.fields["id"]
	var/datum/data/record/security = security_map[id]

	var/list/record = list(
		"id" = id,
		"general_name" = general.fields["name"],
		"general_rank" = general.fields["rank"],
		"general_age" = general.fields["age"],
		"general_sex" = general.fields["sex"],
		"general_m_stat" = general.fields["m_stat"],
		"general_p_stat" = general.fields["p_stat"],
		"security_criminal" = security ? security.fields["criminal"] : "",
		"security_comments" = security ? security.fields["comments"] : null,
		"security_incident" = security ? security.fields["incident"] : null,
		"general_photo_front" = general.fields["photo_front"] ? icon2html(general.fields["photo_front"], user.client, sourceonly=TRUE) :  null,
		"general_photo_side" = general.fields["photo_side"] ? icon2html(general.fields["photo_side"], user.client, sourceonly=TRUE) : null ,
		)
	records += list(record)

.["records"] = records
```

BartDrown replied:
Figured out how to properly pass that in `select_record` act. Now all should be safe and sound, I don't expect any performance issues regarding that. Also forceRefresh of asset is gone, so it should be top notch.

harryob commented on tgui/packages/tgui/interfaces/SecurityRecords.jsx:
```diff
@@ -0,0 +1,688 @@
+          </>
+        )}
+
+        <hr />
```
`<Divider>` component

harryob commented on tgui/packages/tgui/interfaces/SecurityRecords.jsx:
```diff
@@ -0,0 +1,688 @@
+          </Button.Confirm>
+        </Flex>
+
+        <hr />
```
`<Divider>` component

BartDrown replied:
Changed all <hr/> elements to <Divider/>

harryob commented on tgui/packages/tgui/interfaces/SecurityRecords.jsx:
```diff
@@ -0,0 +1,688 @@
+    setNewComment('');
+  };
+
+  const cellStyle = {
+    paddingTop: '4px',
+    paddingBottom: '3px',
+    textAlign: 'center',
+  };
+
+  const boxStyle = {
+    paddingTop: '5px',
+    paddingBottom: '5px',
+  };
+
+  const sectionHeaderStyle = {
+    paddingTop: '10px',
+    paddingBottom: '10px',
+  };
+
+  const grayItalicStyle = {
+    fontSize: '0.9rem',
+    color: 'gray',
+    fontStyle: 'italic',
+  };
```
why in line styles and not in .scss?

BartDrown replied:
Had previously problem with that, but figured it out eventually.
Moved all styles to proper scss file.

BartDrown commented:
Addressed all issues raised in the review, please re-review it. 

I've also added `cursor: pointer` class to TGUI `<Button>` component. 
I couldn't find reason why it shouldn't be like that. Works also pretty good. I think it's good UX change

BartDrown commented on code/game/machinery/computer/security.dm:
```diff
@@ -564,3 +152,404 @@ What a mess.*/
+	//? Leaving below line for debug
+	// to_chat(usr, "Received action [action] with params [json_encode(params)]")
+
+	if(!allowed(usr))
```
Assigned it to variable and reused acros whole call

BartDrown commented on code/game/machinery/computer/security.dm:
```diff
@@ -564,3 +152,404 @@ What a mess.*/
+		return
+
+	//? Leaving below line for debug
+	// to_chat(usr, "Received action [action] with params [json_encode(params)]")
```
Removed debug lines

BartDrown commented on code/game/machinery/computer/security.dm:
```diff
@@ -564,3 +152,404 @@ What a mess.*/
+			var/datum/data/record/general_record = find_record("general", id)
+
+			if(!general_record)
+				alert(usr,"Record not found.")
```
Changed to tgui_alert

BartDrown commented on code/game/machinery/computer/security.dm:
```diff
@@ -564,3 +152,404 @@ What a mess.*/
+
+			var/validation_error = validate_field(field, value, usr, FALSE)
+			if (validation_error)
+				to_chat(usr, "Error: [validation_error]")
```
Changed to tgui_alert

BartDrown replied:
Applied the suggested changes

BartDrown commented on code/game/machinery/computer/security.dm:
```diff
@@ -564,3 +152,404 @@ What a mess.*/
+			var/id = params["id"]
+
+			if(!id)
+				alert(usr,"Invalid record ID.")
```
Changed to tgui_alert

BartDrown commented on code/game/machinery/computer/security.dm:
```diff
@@ -564,3 +152,404 @@ What a mess.*/
+							H.sec_hud_set_security_status()
+
+				else
+					alert(usr, "Field not found in security record.")
```
Applied the suggested changes

BartDrown commented on code/game/machinery/computer/security.dm:
```diff
@@ -564,3 +152,404 @@ What a mess.*/
+					alert(usr, "Field not found in security record.")
+					return
+			else
+				alert(usr, "Record not found or mismatched field type.")
```
Applied the suggested changes

BartDrown commented on code/game/machinery/computer/security.dm:
```diff
@@ -564,3 +152,404 @@ What a mess.*/
+			else
+				security_record.fields["comments"]["[comment_id]"] = new_comment
+
+			to_chat(usr, "Comment added successfully.")
```
Applied the suggested changes

BartDrown commented on code/game/machinery/computer/security.dm:
```diff
@@ -564,3 +152,404 @@ What a mess.*/
+
+		if ("clear_fingerprints")
+			if (!scanner)
+				to_chat(usr, "No scanner found.")
```
Applied the suggested changes

BartDrown commented on code/game/machinery/computer/security.dm:
```diff
@@ -564,3 +152,404 @@ What a mess.*/
+			var/photo_profile = params["photo_profile"]
+			var/icon/img = get_photo(usr)
+			if(!img)
+				to_chat(usr, "You are currently not holding any photo")
```
Applied the suggested changes

BartDrown commented on code/game/machinery/computer/security.dm:
```diff
@@ -564,3 +152,404 @@ What a mess.*/
+		//* Actions for ingame objects interactions
+		if("print_fingerprint_report")
+			if (!( printing ))
+				printing = 1
```
Applied the suggested changes

BartDrown commented on code/game/machinery/computer/security.dm:
```diff
@@ -564,3 +152,404 @@ What a mess.*/
+			if (!( printing ))
+				printing = 1
+				if (!scanner || !scanner.print_list)
+					to_chat(usr, "No scanner data found.")
```
Applied the suggested changes

BartDrown commented on code/game/machinery/computer/security.dm:
```diff
@@ -564,3 +152,404 @@ What a mess.*/
+				return
+
+			scanner.update_icon()
+			to_chat(usr, "Fingerprints cleared from the scanner.")
```
Applied the suggested changes

BartDrown commented on code/game/machinery/computer/security.dm:
```diff
@@ -564,3 +152,404 @@ What a mess.*/
+			// Locate the general record
+			var/datum/data/record/general_record = find_record("general", id)
+
+			if (!general_record)
```
Applied the suggested changes

BartDrown commented on code/game/machinery/computer/security.dm:
```diff
@@ -564,3 +152,404 @@ What a mess.*/
+				else
+					alert(usr, "Field not found in security record.")
+					return
+			else
```
Updated it to this little monster:
```dm
// Update the appropriate record
if(is_general_field && general_record && (original_field in general_record.fields))
	general_record.fields[original_field] = value

else if(is_security_field && security_record && (original_field in security_record.fields))
	security_record.fields[original_field] = value

	if(original_field == "criminal")
		for(var/mob/living/carbon/human/H in GLOB.human_mob_list)
			H.sec_hud_set_security_status()
else
	tgui_alert(user, "Record or associated field not found.")
	return
```
Now 50% less code, with greatly reduced readability!

BartDrown commented:
Addressed all issues raised in the review, please re-review it.

Also added some additional sounds when using machine.
